### PR TITLE
Correct JSON Unmarshaling for Text and Image Content

### PR DESCRIPTION
**Pull request by nvtrinh2001**
## Background 

- Issue: #79 
- This PR removes incorrect logic of `Content.UnmarshalJSON` and improves parsing logic for `Part`, `TextContent`, and `ImageContent`, ensuring better handling of different content types, extra fields, and invalid inputs. It also introduces comprehensive test cases to validate the behavior.

## Changes 

- Implemented a structured approach for unmarshaling JSON into `Part`, `TextContent`, and `ImageContent` to prevent errors from extra unknown fields.
- Added validation checks for empty or missing required fields (e.g., empty `text` in `TextContent` or missing `url` in `ImageContent`).
- Enhanced error handling for invalid JSON structures and unknown content types.
- Introduced tests for:
  - Valid and invalid `text` and `image` content.
  - Extra unknown fields in JSON inputs.
  - Handling of incorrect types or malformed JSON structures.

## Considerations 

- Verify the correctness of JSON parsing and error handling logic.
- Ensure all test cases cover edge cases effectively.

## Remaining Tasks

### CODE REVIEW

File: openai/openai_test.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+			input: `[]`,
+			expected: MessageContent{
+				String: nil,
+				Parts:  []Part{},
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (abdulhamidnugroho) [inline openai/openai_test.go]: As I was told in another [PR](https://github.com/yanolja/ogem/pull/65#discussion_r2007168719)
> should not use any conditional statements like if, for, switch, etc in tests

cc: @seungduk-yanolja

Comment (abdulhamidnugroho) [reply openai/openai_test.go]: Sorry, just found out it has already been refactored

File: openai/openai.go [hunk 2]
```diff
@@ -215,14 +215,30 @@ func (p *Content) MarshalJSON() ([]byte, error) {
...
+		if textStr, ok := text.(string); ok && textStr != "" {
+			p.TextContent = &TextContent{Text: textStr}
+			return nil
+		}
+		return fmt.Errorf("expected text content, got %v", text)
 	}
-	var image ImageContent
-	if err := json.Unmarshal(data, &image); err == nil {
-		p.ImageContent = &image
+	if url, ok := raw["url"]; ok {
```
Comment (seungduk-yanolja) [inline openai/openai.go:229]: What if both exist?

Comment (nvtrinh2001) [reply openai/openai.go]: Really good point. Logically, they shouldn't exist at the same time since in `Part` definition, there is `type` field for the content:
```
type Part struct {
	Type    string  `json:"type"`
	Content Content `json:"content"`
}

type Content struct {
	TextContent  *TextContent
	ImageContent *ImageContent
}
```
However, if these fields both exist for some reasons, my `UnmarshalJSON` method will return value of the text, and nil for the image content. For example with this input:
```
input := `{
	"text": "hello",
	"url": "https://example.com/image.jpg", 
	"detail": "high"	
}`
```
The method should return the content based on the type, but the actual output will always be:
```
expected := Content{
	TextContent: &TextContent{Text: "hello"},
	ImageContent: nil,
}
```
To solve this, I think the best approach is to update the `UnmarshalJSON` method of `Part` struct. Based on the `type` field of the struct, we can decide to keep the `TextContent` or the `ImageContent`. 

I updated the change to this PR.
